### PR TITLE
Adding reserve() to VertexArray

### PR DESCRIPTION
**Pull request by HerrNamenlos123**
* [ ] Has this change been discussed on [the forum](https://en.sfml-dev.org/forums/index.php#c3) or in an issue before?
* [x] Does the code follow the SFML [Code Style Guide](https://www.sfml-dev.org/style.php)?
* [ ] Have you provided some example/test code for your changes?
* [ ] If you have additional steps which need to be performed list them as tasks!

----

## Description

Hi there, thank you so much for this amazing library!
I am trying to implement a Batch Renderer and therefore need all the performance I can get. I added a small change to also expose the `reserve()` function besides the `resize()` function. Any problems with that? :)

I also added a unit test, although we cannot really check if the allocation works, we can only check that the function is callable.

## Tasks

* [ ] Tested on Linux
* [ ] Tested on Windows
* [ ] Tested on macOS
* [ ] Tested on iOS
* [ ] Tested on Android

## How to test this PR?

```cpp
void resize() {
    sf::VertexArray va;
    va.reserve(50000);       // Thanks to this function all 50000 elements are dynamically appended, with only a single heap allocation
    for (int i = 0; i < 50000; i++) {
        va.append(sf::Vertex(sf::Vector2f(0, 0), sf::Color::Red));
    }
}
```

### CODE REVIEW

**Comment by Marioalexsan** [comment]
There's a [pull request](https://github.com/SFML/SFML/issues/2208) related to VertexArrays that aims to outright expose the internal array, since there is little point in encapsulating it. Could you look at this too and see whenever this could be more useful?

**Comment by HerrNamenlos123** [comment]
Thank you, didn't know the discussion exists. If that PR gets resolved at some point, my PR is irrelevant. I think making it a struct or removing it completely is the best option.

**Comment by Bromeon** [comment]
See also related discussion in the past in `sf::Packet`: https://github.com/SFML/SFML/issues/1752#issuecomment-820194385

**Comment by danieljpetersen** [comment]
Does this have to be blocked on that other issue? This seems like a very uncontroversial change and purely a net benefit. There is pushback and no consensus on turning the vertex array into a plain struct, I can imagine that proposal getting stuck in limbo for a long time (and indeed, it already has been a while). Seems like the preferable course would be to merge this in and take the win now, and if later it's decided to do something more drastic this isn't harming that effort.

**Comment by Bambo-Borris** [comment]
> Does this have to be blocked on that other issue? This seems like a very uncontroversial change and purely a net benefit. There is pushback and no consensus on turning the vertex array into a plain struct, I can imagine that proposal getting stuck in limbo for a long time (and indeed, it already has been a while). Seems like the preferable course would be to merge this in and take the win now, and if later it's decided to do something more drastic this isn't harming that effort.


I agree about the concerns of it being a great idea lost if blocked by the struct one, but also the struct one would entirely undo the need for this. I think I'd rather see someone open a prospective PR for the struct change and have it championed. I took a look, and there's lots of discussion on the issue, but I didn't see any link to a potential implementation. An actual PR for the struct issue would have a fighting chance. Just remaining an issue will lead to never-ending conversation.

File: include/SFML/Graphics/VertexArray.hpp
Comment (ChrisThrasher) [inline include/SFML/Graphics/VertexArray.hpp:129]:
```suggestion
    /// \param capacity New capacity of the array (number of vertices)
    ///
    ////////////////////////////////////////////////////////////
    void reserve(std::size_t capacity);
```

File: src/SFML/Graphics/VertexArray.cpp
Comment (ChrisThrasher) [inline src/SFML/Graphics/VertexArray.cpp:75]:
```suggestion
void VertexArray::reserve(std::size_t capacity)
{
    m_vertices.reserve(capacity);
```

**Review by ChrisThrasher — CHANGES_REQUESTED** [review]
I think I'm okay with this PR, but it comes with the caveat that we still may refactor `sf::VertexArray` into an aggregate type or something else different from its current form.

**Comment by ChrisThrasher** [comment]
https://github.com/SFML/SFML/compare/master...ChrisThrasher:SFML:vertex_array

I reimplemted `sf::VertexArray` to have all public data and it has some nice upsides. We delete 200 lines of code and can now leverage `std::vector` features that were previously inaccessible.

**Comment by ChrisThrasher** [comment]
@eXpl0it3r Do you want to close this based on your comments [here](https://github.com/SFML/SFML/issues/2208#issuecomment-1606299436)?

**Comment by shane-tran-whitmire** [comment]
I mean, as long as there is a way to do this highly custom functionality (there is if you just make and manage your own std::vector<sfml::Vertex>), then i think it is fine not to have this function.

**Comment by ChrisThrasher** [comment]
> I mean, as long as there is a way to do this highly custom functionality

I believe so but I have yet to someone illustrate that.

https://github.com/SFML/SFML/blob/11b73743c42cf7ecd7c596ba83fdbf1150ffa96c/include/SFML/Graphics/RenderTarget.hpp#L248-L249

The above function (that gets inherited by `sf::RenderWindow`) seems to be how you're supposed to draw a vector of vertices without using `sf::VertexArray`. I believe this is what you'll be forced to use if you want to use any `std::vector` interfaces that `sf::VertexArray` doesn't already expose.

**Comment by ChrisThrasher** [comment]
I personally wish we could merge this feature but there isn't consensus from the other maintainers so I'm going to close this.

**Pull request closed without merging**
